docs/guide/api-vite-environment.md

```diff
@@ -0,0 +1,874 @@
+    send: (data) => connection.send(data),
+    onMessage: (listener) => connection.on('message', listener),
+  })
+
+  const workerdDevEnvironment = new DevEnvironment(name, config, {
+    options: {
+      resolve: { conditions: ['custom'] },
+      ...options,
+    },
+    hot,
+    runner: {
+      transport,
+    },
+  })
+  return workerdDevEnvironment
+}
+```
+
+Then users can create a workerd environment to do SSR using:
+
+```js
+const ssrEnvironment = createWorkerdEnvironment('ssr', config)
+```
+
+## Environment Configuration
+
+Environments are explicitely configured with the `environments` config option.
+
+```js
+export default {
+  environments: {
+    client: {
+      resolve: {
+        conditions: [], // configure the Client environment
+      },
+    },
+    ssr: {
+      dev: {
+        optimizeDeps: {}, // configure the SSR environment
+      },
+    },
+    rsc: {
+      resolve: {
+        noExternal: true, // configure a custom environment
+      },
+    },
+  },
+}
+```
+
+Vite's user config also extends from a environment config, letting users add defaults for all environments at the root level. This is quite useful for the common use case of configuring a Vite client only app, that can be done without going through `environments.client`.
+
+```js
+export default {
+  resolve: {
+    conditions: [], // configure a default for all environments
+  },
+}
+```
+
+The `EnvironmentOptions` interface exposes all the per-environment options. There are `SharedEnvironmentOptions` that apply to both `build` and `dev` environments, like `resolve`. And there are `DevOptions` and `BuildOptions`
+
+```ts
+interface EnvironmentOptions extends SharedEnvironmentOptions {
+  dev: DevOptions
+  build: BuildOptions
+}
+```
+
+As we explained, the `UserConfig` interface extends from `EnvironmentConfig`. Environment specific options defined at the root level of user config are used for the default client environment. And environments can be configured explicitely using the `environments` array. The Client and SSR environments, are always present, even if an empty object is set to `environments`.
+
+```ts
+interface UserConfig extends EnvironmentOptions {
+  environments: Record<string, EnvironmentOptions>
+  // other options
+}
+```
+
+::: info
+
+The `ssr` top level property has many options in common with `EnvironmentOptions`. This option was created for the same use case as `environments` but only allowed configuration of a small number of options. We're going to deprecate it in favour of a unified way to define environment configuration.
+
+:::
+
+## Custom environment instances
+
+To register a new dev or build environment, you can use a `create` function:
+
+```js
+export default {
+  environments: {
+    rsc: {
+      dev: {
+        createEnvironment(name, config) {
+          return createNodeDevEnvironment(name, config)
+        }
+      },
+      build: {
+        createEnvironment(name, config) {
```

Comment:
Hopefully with the added context at the top is more clear. There is a single environment, and we need to configure how it should be built and how the it works during dev. Some environments may target a different environment during dev because it is hard to emulate locally (hopefully this isn't going to be as common as it was in the past)